File: src/confined-files.py
"""Bounded one-request POSIX filesystem helper, launched with Python -I -S."""
import base64
import errno
import json
import os
import secrets
import stat
import sys

MAX_FILE = 2 * 1024 * 1024
MAX_REQUEST = 3 * 1024 * 1024
MAX_RESPONSE = 8 * 1024 * 1024
MAX_ENTRIES = 30000
DIRECTORY = os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW | os.O_CLOEXEC


def validate(root, relative, directory=False):
    if (not isinstance(root, str) or not root.startswith('/') or root == '/'
            or '\x00' in root or any(p in ('', '.', '..') for p in root[1:].split('/'))):
        raise ValueError('Unsafe filesystem root')
    if directory and relative == '':
        return
    if (not isinstance(relative, str) or not relative
            or any(c in relative for c in '\\*?[]{}')
            or any(ord(c) < 32 for c in relative)
            or any(p in ('', '.', '..') for p in relative.split('/'))):
        raise ValueError('Unsafe relative path')


def directory_fd(root, relative='', create=False):
    fd = os.open('/', DIRECTORY)
    try:
        root_parts = root[1:].split('/')
        parts = root_parts + (relative.split('/') if relative else [])
        for index, part in enumerate(parts):
            if create and index >= len(root_parts):
                try:
                    os.mkdir(part, mode=0o700, dir_fd=fd)
                except FileExistsError:
                    pass
            next_fd = os.open(part, DIRECTORY, dir_fd=fd)
            os.close(fd)
            fd = next_fd
        return fd
    except BaseException:
        os.close(fd)
        raise


def safe_file(info):
    if not stat.S_ISREG(info.st_mode) or info.st_nlink != 1:
        raise ValueError('Unsafe file')
    if info.st_size > MAX_FILE:
        raise ValueError('Oversized file')


def destination(fd, name):
    try:
        safe_file(os.stat(name, dir_fd=fd, follow_symlinks=False))
    except FileNotFoundError:
        pass


def operate(request):
    operation = request.get('operation')
    root, relative = request.get('root'), request.get('relative')
    validate(root, relative, operation == 'entries')
    if operation == 'entries':
        fd = directory_fd(root, relative)
        try:
            entries, size = [], 0
            with os.scandir(fd) as iterator:
                for entry in iterator:
                    kind = 'directory' if entry.is_dir(follow_symlinks=False) else 'file' if entry.is_file(follow_symlinks=False) else 'other'
                    item = {'name': entry.name, 'type': kind}
                    size += len(json.dumps(item, ensure_ascii=True)) + 1
                    if len(entries) >= MAX_ENTRIES or size > MAX_RESPONSE - 1024:
                        raise ValueError('Workspace entry limit exceeded')
                    entries.append(item)
            return entries
        finally:
            os.close(fd)
    if operation not in ('read', 'write', 'remove'):
        raise ValueError('Unsafe filesystem operation')
    data, mode = None, request.get('mode', 0o600)
    if operation == 'write':
        if not isinstance(mode, int) or isinstance(mode, bool) or not 0 <= mode <= 0o777:
            raise ValueError('Unsafe file mode')
        data = base64.b64decode(request.get('data', ''), validate=True)
        if len(data) > MAX_FILE:
            raise ValueError('Oversized file')
    parent, _, name = relative.rpartition('/')
    fd = directory_fd(root, parent, operation == 'write')
    try:
        if operation == 'read':
            file_fd = os.open(name, os.O_RDONLY | os.O_NOFOLLOW | os.O_NONBLOCK | os.O_CLOEXEC, dir_fd=fd)
            try:
                info = os.fstat(file_fd)
                safe_file(info)
                chunks, length = [], 0
                while length <= MAX_FILE:
                    chunk = os.read(file_fd, min(65536, MAX_FILE + 1 - length))
                    if not chunk:
                        break
                    chunks.append(chunk)
                    length += len(chunk)
                safe_file(os.fstat(file_fd))
                if length > MAX_FILE:
                    raise ValueError('Oversized file')
                return {'data': base64.b64encode(b''.join(chunks)).decode('ascii'), 'mode': info.st_mode & 0o777}
            finally:
                os.close(file_fd)
        if operation == 'remove':
            safe_file(os.stat(name, dir_fd=fd, follow_symlinks=False))
            os.unlink(name, dir_fd=fd)
            return None
        destination(fd, name)
        temporary = '.privacy-' + secrets.token_hex(16)
        created = False
        try:
            file_fd = os.open(temporary, os.O_WRONLY | os.O_CREAT | os.O_EXCL | os.O_NOFOLLOW | os.O_CLOEXEC, mode, dir_fd=fd)
            created = True
            try:
                remaining = memoryview(data)
                while remaining:
                    count = os.write(file_fd, remaining)
                    if count <= 0:
                        raise OSError(errno.EIO, 'File write made no progress')
                    remaining = remaining[count:]
            finally:
                os.close(file_fd)
            destination(fd, name)
            os.replace(temporary, name, src_dir_fd=fd, dst_dir_fd=fd)
        finally:
            if created:
                try:
                    os.unlink(temporary, dir_fd=fd)
                except FileNotFoundError:
                    pass
        return None
    finally:
        os.close(fd)


def main():
    try:
        data = sys.stdin.buffer.read(MAX_REQUEST + 1)
        if len(data) > MAX_REQUEST:
            raise ValueError('Filesystem request limit exceeded')
        request = json.loads(data)
        if not isinstance(request, dict):
            raise ValueError('Invalid filesystem request')
        response = {'ok': True, 'result': operate(request)}
    except Exception as error:
        response = {'ok': False, 'error': str(error)[:2048], 'code': errno.errorcode.get(error.errno) if isinstance(error, OSError) else None}
    encoded = json.dumps(response, ensure_ascii=True).encode('ascii')
    if len(encoded) > MAX_RESPONSE:
        encoded = b'{"ok":false,"error":"Filesystem response limit exceeded"}'
    sys.stdout.buffer.write(encoded)


if __name__ == '__main__':
    main()

File: src/confined-files.js
import fs from 'node:fs/promises';
import { constants } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { randomUUID } from 'node:crypto';
import { spawn } from 'node:child_process';

const MAX_FILE = 2 * 1024 * 1024;
const MAX_ENTRIES = 30000;
const MAX_RESPONSE = 8 * 1024 * 1024;
const helper = fileURLToPath(new URL('./confined-files.py', import.meta.url));
function validate(root, relative, directory = false) {
  if (typeof root !== 'string' || !path.isAbsolute(root) || root === '/' || path.resolve(root) !== root || root.includes('\0')) throw new Error('Unsafe filesystem root');
  if (directory && relative === '') return;
  if (typeof relative !== 'string' || !relative || relative.includes('\\') || /[\x00-\x1f*?\[\]{}]/.test(relative) || relative.split('/').some(p => !p || p === '.' || p === '..')) throw new Error('Unsafe relative path');
}
function content(data, mode) {
  if (typeof data !== 'string' && !Buffer.isBuffer(data)) throw new Error('Invalid file content');
  const buffer = Buffer.isBuffer(data) ? data : Buffer.from(data);
  if (buffer.length > MAX_FILE) throw new Error('Oversized file');
  if (!Number.isInteger(mode) || mode < 0 || mode > 0o777) throw new Error('Unsafe file mode');
  return buffer;
}
function safeFile(stat) {
  if (!stat.isFile() || stat.nlink !== 1) throw new Error('Unsafe file');
  if (stat.size > MAX_FILE) throw new Error('Oversized file');
}
// Start at / so O_NOFOLLOW protects the root's ancestors as well as its leaf.
async function directoryHandle(root, relative = '', create = false) {
  let handle = await fs.open('/', constants.O_RDONLY | constants.O_DIRECTORY | constants.O_NOFOLLOW);
  try {
    const rootParts = root.slice(1).split('/');
    const parts = [...rootParts, ...(relative ? relative.split('/') : [])];
    for (let i = 0; i < parts.length; i++) {
      const child = `/proc/self/fd/${handle.fd}/${parts[i]}`;
      if (create && i >= rootParts.length) await fs.mkdir(child, { mode: 0o700 }).catch(e => { if (e.code !== 'EEXIST') throw e; });
      const next = await fs.open(child, constants.O_RDONLY | constants.O_DIRECTORY | constants.O_NOFOLLOW);
      await handle.close(); handle = next;
    }
    return handle;
  } catch (error) { await handle.close(); throw error; }
}
async function parentHandle(root, relative, create = false) {
  const parts = relative.split('/'); const name = parts.pop();
  const handle = await directoryHandle(root, parts.join('/'), create);
  return { handle, filename: `/proc/self/fd/${handle.fd}/${name}` };
}
function linuxFiles() {
  return {
    async read(root, relative) {
      validate(root, relative); const parent = await parentHandle(root, relative); let file;
      try {
        file = await fs.open(parent.filename, constants.O_RDONLY | constants.O_NOFOLLOW | constants.O_NONBLOCK);
        const stat = await file.stat(); safeFile(stat);
        const buffer = Buffer.alloc(MAX_FILE + 1); let length = 0;
        while (length < buffer.length) {
          const { bytesRead } = await file.read(buffer, length, buffer.length - length, null);
          if (!bytesRead) break;
          length += bytesRead;
        }
        safeFile(await file.stat());
        if (length > MAX_FILE) throw new Error('Oversized file');
        return { data: Buffer.from(buffer.subarray(0, length)), mode: stat.mode & 0o777 };
      } finally { await file?.close(); await parent.handle.close(); }
    },
    async write(root, relative, data, mode = 0o600) {
      validate(root, relative); const buffer = content(data, mode);
      const parent = await parentHandle(root, relative, true);
      const temp = `${parent.filename}.privacy-${randomUUID()}`;
      const checkDestination = async () => { try { safeFile(await fs.lstat(parent.filename)); } catch (e) { if (e.code !== 'ENOENT') throw e; } };
      try {
        await checkDestination();
        await fs.writeFile(temp, buffer, { flag: 'wx', mode });
        await checkDestination();
        await fs.rename(temp, parent.filename);
      } finally { await fs.unlink(temp).catch(() => {}); await parent.handle.close(); }
    },
    async remove(root, relative) {
      validate(root, relative); const parent = await parentHandle(root, relative);
      try { safeFile(await fs.lstat(parent.filename)); await fs.unlink(parent.filename); }
      finally { await parent.handle.close(); }
    },
    async entries(root, relative = '') {
      validate(root, relative, true); const handle = await directoryHandle(root, relative); let directory;
      try {
        directory = await fs.opendir(`/proc/self/fd/${handle.fd}`);
        const entries = []; let bytes = 0;
        for await (const entry of directory) {
          const item = { name: entry.name, type: entry.isDirectory() ? 'directory' : entry.isFile() ? 'file' : 'other' };
          bytes += Buffer.byteLength(JSON.stringify(item)) + 1;
          if (entries.length >= MAX_ENTRIES || bytes > MAX_RESPONSE - 1024) throw new Error('Workspace entry limit exceeded');
          entries.push(item);
        }
        return entries;
      } finally { await directory?.close().catch(() => {}); await handle.close(); }
    },
  };
}
function pythonFiles(pythonPath) {
  if (typeof pythonPath !== 'string' || pythonPath && (!path.isAbsolute(pythonPath) || pythonPath.includes('\0'))) throw new Error('Python path must be absolute');
  let interpreter;
  async function resolveInterpreter() {
    if (pythonPath) return pythonPath;
    if (!interpreter) interpreter = (async () => {
      for (const candidate of ['/usr/bin/python3', '/opt/homebrew/bin/python3', '/usr/local/bin/python3']) {
        try { await fs.access(candidate, constants.X_OK); return candidate; } catch {}
      }
      throw new Error('Python 3 unavailable; configure an absolute pythonPath');
    })();
    return interpreter;
  }
  async function invoke(operation, root, relative, extra = {}) {
    validate(root, relative, operation === 'entries');
    const input = JSON.stringify({ operation, root, relative, ...extra });
    if (Buffer.byteLength(input) > 3 * 1024 * 1024) throw new Error('Filesystem request limit exceeded');
    const executable = await resolveInterpreter();
    return new Promise((resolve, reject) => {
      const child = spawn(executable, ['-I', '-S', helper], { cwd: '/', env: { PATH: '/usr/bin:/bin', LANG: 'C.UTF-8' }, stdio: ['pipe', 'pipe', 'pipe'] });
      const chunks = []; let size = 0; let error;
      const stop = reason => { error ??= reason; child.kill('SIGKILL'); };
      const timer = setTimeout(() => stop(new Error('Confined filesystem helper timeout')), 15000);
      child.stdout.on('data', data => { size += data.length; if (size > MAX_RESPONSE) stop(new Error('Filesystem response limit exceeded')); else chunks.push(data); });
      let stderrBytes = 0;
      child.stderr.on('data', data => { stderrBytes += data.length; if (stderrBytes > 16384) stop(new Error('Filesystem helper error output limit exceeded')); });
      child.on('error', cause => { error = new Error('Python 3 unavailable for confined filesystem', { cause }); });
      child.stdin.on('error', cause => { if (cause.code !== 'EPIPE') stop(cause); });
      child.on('close', code => {
        clearTimeout(timer);
        if (error) return reject(error);
        try {
          const result = JSON.parse(Buffer.concat(chunks).toString('utf8'));
          if (!result.ok) throw Object.assign(new Error(result.error || 'Confined filesystem failed'), { code: result.code });
          if (code !== 0) throw new Error('Confined filesystem helper failed');
          resolve(result.result);
        } catch (cause) { reject(cause); }
      });
      child.stdin.end(input);
    });
  }
  return {
    async read(root, relative) { const result = await invoke('read', root, relative); return { data: Buffer.from(result.data, 'base64'), mode: result.mode }; },
    async write(root, relative, data, mode = 0o600) { const buffer = content(data, mode); await invoke('write', root, relative, { data: buffer.toString('base64'), mode }); },
    async remove(root, relative) { await invoke('remove', root, relative); },
    async entries(root, relative = '') { return invoke('entries', root, relative); },
  };
}
export function createConfinedFiles({ backend = 'auto', pythonPath = '' } = {}) {
  const selected = backend === 'auto' ? process.platform === 'linux' ? 'linux' : 'python' : backend;
  if (selected === 'python') return pythonFiles(pythonPath);
  if (selected === 'linux' && process.platform === 'linux') return linuxFiles();
  throw new Error('Unsupported confined filesystem backend');
}
